File: java/695-max-area-of-island.java
class Solution {
    private int ans = 0;
    class Coordinate {
        int x;
        int y;
        public Coordinate(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    public int maxAreaOfIsland(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1 && !visited[i][j]) {
                    bfs(grid, i, j, visited);
                }
            }
        }
        return ans;
    }
    private void bfs(int[][] grid, int x, int y, boolean[][] visited) {
        int[] directionX = {0, 0, 1, -1};
        int[] directionY = {1, -1, 0, 0};
        Queue<Coordinate> queue = new LinkedList<>();
        queue.offer(new Coordinate(x, y));
        visited[x][y] = true;
        int num = 0;
        while (!queue.isEmpty()) {
            Coordinate curr = queue.poll();
            // System.out.println(String.format("x=%d,y=%d", curr.x, curr.y));
            num++;
            for (int i = 0; i < 4; i++) {
                int nextX = curr.x + directionX[i];
                int nextY = curr.y + directionY[i];
                // System.out.println(String.format("nextX=%d,nextY=%d", nextX, nextY));
                if (isValid(grid, nextX, nextY) && !visited[nextX][nextY] && grid[nextX][nextY] == 1) {
                    queue.offer(new Coordinate(nextX, nextY));
                    visited[nextX][nextY] = true;
                }
            } 
        }
        // System.out.println(num);
        ans = Math.max(ans, num);
        
    }
    private boolean isValid(int[][] grid, int x, int y) {
        return 0 <= x && x < grid.length && 0 <= y && y < grid[0].length;
    }
}

File: python3/695.max-area-of-island.py
DIRECTIONS = (
    (0, 1),
    (0, -1),
    (1, 0),
    (-1, 0)
)

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        
        
        ans = 0
        
        def flood(x, y):
            grid[x][y] = 0
            area = 1
            for diff_x, diff_y in DIRECTIONS:
                next_x = x + diff_x
                next_y = y + diff_y
                if in_bound(next_x, next_y, n, m) and grid[next_x][next_y] == 1:
                    area += flood(next_x, next_y)
            return area
            
        
        for x in range(n):
            for y in range(m):
                if grid[x][y] == 1:
                    ans = max(ans, flood(x, y))
        return ans
    
def in_bound(x, y, n, m):
    return 0 <= x < n and 0 <= y < m
